hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/enrichment/Enricher.java

```diff
@@ -21,6 +22,9 @@ public interface Enricher {
   /** Enrich the attributes/metrics for an Event */
   void enrichEvent(StructuredTrace trace, Event event);
 
+  /** Enrich the attributes/metrics for an Event with support for metrics */
+  void enrichEvent(StructuredTrace trace, Event event, Counter errorCounter);
```

Comment:
I didn't follow how this was fully tied. But, not sure this is the way we should move forward.

Here, we are changing the interface of `enrichEvent` for a specific metric type - `Counter`.
In future, if we need a different metric say timer, we need a new interface.

To me, it likes that all enricher should have access to global metric registry.
So, it can modify its counter accordingly.

Does above has some issue with multiple stream thread? Is that the reason we are going ahead with interface based approach?

Reply:
The idea is to use reuse the already created traceErrorCounter here:

https://github.com/hypertrace/hypertrace-ingester/pull/400/files#diff-a94acb212965787a7cc6f9e663b0cb9071efcb7f3024d3a44ba2de50d9063c27L99

Intention is to have a way to count the number of exceptions that were thrown in an enricher which got handled.

Uncaught exceptions will be counted by the same counter in the above code but currently to count the exceptions which were handled inside the enricher itself is not possible. 

We want to solve that problem by this interface. We can discuss if there is a better to to achieve this

Reply:
The overloading of existing `enrichEvent()` to pass Counter argument doesn't look clean for multiple reasons:
- It is coupled with a particular metric impl (codahale today; something else later), which is more of an implementation detail.
- Also tomorrow if some other metric (instead of Counter) is needed, we can't keep on changing method signature.
- Emitting metrics is more of an implementation aspect (not exactly a functional requirement of enrichment).

As @kotharironak mentioned, a cleaner approach is to have all enrichers access to the the global metric registry, from where they can directly access the desired Counter and then increment it.

Reply:
@satish-mittal had a discussion offline with @kotharironak , will use a global metric registry.